src/main/java/com/odysseusinc/arachne/portal/api/v1/controller/BaseAnalysisController.java

```diff
@@ -444,30 +460,56 @@ public JsonResult updateCommonEntityInAnalysis(@PathVariable("analysisId") Long
     @ApiOperation("Upload file to attach to analysis.")
     @RequestMapping(value = "/api/v1/analysis-management/analyses/{analysisId}/upload", method = POST)
     public JsonResult<List<AnalysisFileDTO>> uploadFile(Principal principal,
-                                                        @Valid UploadFileDTO uploadFileDTO,
+                                                        @Valid UploadFilesDTO uploadFilesLinks,
                                                         @PathVariable("analysisId") Long id)
-            throws PermissionDeniedException, NotExistException, IOException {
+            throws PermissionDeniedException, NotExistException {
 
         IUser user = getUser(principal);
         T analysis = analysisService.getById(id);
-
         JsonResult<List<AnalysisFileDTO>> result = new JsonResult<>(NO_ERROR);
-
         List<AnalysisFileDTO> createdFiles = new ArrayList<>();
+        Map<String, Object> validationErrors = new HashMap<>();
+        List<String> errorFileMessages = new ArrayList<>();
+        uploadFilesLinks.getFiles().forEach(uploadFile -> {
+            if (uploadFile.getFile() != null) {
+                AnalysisFile createdFile = null;
+                try {
+                    createdFile = analysisService.saveFile(uploadFile.getFile(), user, analysis, uploadFile.getLabel(), uploadFile.getExecutable(), null);
+                } catch (IOException e) {
+                    LOGGER.error("Failed to save file", e);
+                } catch (AlreadyExistException e) {
+                    LOGGER.error("Failed to save file", e);
+                    result.setErrorCode(ALREADY_EXIST.getCode());
+                    errorFileMessages.add(e.getMessage());
+                }
+                createdFiles.add(conversionService.convert(createdFile, AnalysisFileDTO.class));
+            } else {
+                result.setErrorCode(VALIDATION_ERROR.getCode());
+            }
+        });
 
-        if (uploadFileDTO.getFile() != null) {
-            AnalysisFile createdFile = analysisService.saveFile(uploadFileDTO.getFile(), user, analysis, uploadFileDTO.getLabel(), uploadFileDTO.getExecutable(), null);
-            createdFiles.add(conversionService.convert(createdFile, AnalysisFileDTO.class));
-        } else {
-            if (StringUtils.hasText(uploadFileDTO.getLink())) {
-                AnalysisFile createdFile = analysisService.saveFile(uploadFileDTO.getLink(), user, analysis, uploadFileDTO.getLabel(), uploadFileDTO.getExecutable());
+        uploadFilesLinks.getLinks().forEach(uploadLink -> {
+            if (StringUtils.hasText(uploadLink.getLink())) {
+                AnalysisFile createdFile = null;
+                try {
+                    createdFile = analysisService.saveFile(uploadLink.getLink(), user, analysis, uploadLink.getLabel(), uploadLink.getExecutable());
+                } catch (IOException e) {
```

Comment:
Looks very repetitive. Cannot we either move exception handling into some wrapper (so pass "saveFile" as lambda to a function, which wraps it with the error handlers) or use single forEach with internal logic for selecting which method to use (e.g. `if (StringUtils.hasText(uploadLink.getLink())) then uploadLink else uploadFile`)